crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -1188,16 +1226,83 @@ impl<'db> Parameters<'db> {
                 .map(|default| definition_expression_type(db, definition, default))
         };
 
+        let method_info = infer_method_information(db, definition);
+        let is_static_or_classmethod = method_info
+            .is_some_and(|f| f.method.is_staticmethod(db) || f.method.is_classmethod(db));
+
+        let inferred_annotation = |arg: &ParameterWithDefault| {
+            if let Some(MethodInformation { method, class }) = method_info
+                && !is_static_or_classmethod
+                && arg.parameter.annotation().is_none()
+                && parameters.index(arg.name().id()) == Some(0)
+                // TODO: find out why we break protocol type property tests when we include them here:
+                && !class.is_protocol(db)
```

Comment:
Will talk to @AlexWaygood about this. I assume that we break assignability/subtyping for protocol since we create `Self` with a bound that refers to the protocol class itself (instead of doing something that is more appropriate for structural types… like having no bound?).